### PR TITLE
feat: index referrers on manifest push

### LINKED ISSUE #312: Add client-side compatibility for pushing manifest with `subject` field

### PR DESCRIPTION
**Pull request by Wwwsylvia**
Resolves #312 
Signed-off-by: Lixia (Sylvia) Lei <lixlei@microsoft.com>

### CODE REVIEW

**Comment by Wwwsylvia** [comment]
Tested against https://github.com/oci-playground/distribution

File: registry/remote/repository.go [hunk 1]
```diff
@@ -1051,6 +1090,113 @@ func (s *manifestStore) push(ctx context.Context, expected ocispec.Descriptor, c
...
+	s.repo.lockReferrersTag(referrersTag)
+	defer s.repo.unlockReferrersTag(referrersTag)
+
+	var updatedReferrers []ocispec.Descriptor
+	err = s.repo.referrersByTagSchema(ctx, subject, "", func(referrers []ocispec.Descriptor) error {
+		for _, r := range referrers {
+			// skip duplicate entry
+			if !content.Equal(r, desc) {
+				updatedReferrers = append(updatedReferrers, r)
+			}
```
Comment (shizhMSFT) [inline registry/remote/repository.go:1147]: If there is a duplicate, does that imply we don't need to update the index?

What if I push `v1.0.0` and then push `v1.0`, `v1`, `latest` with the same manifest?

File: registry/remote/repository.go [hunk 2]
```diff
@@ -878,7 +901,15 @@ func (s *manifestStore) Fetch(ctx context.Context, target ocispec.Descriptor) (r
 
 // Push pushes the content, matching the expected descriptor.
 func (s *manifestStore) Push(ctx context.Context, expected ocispec.Descriptor, content io.Reader) error {
-	return s.push(ctx, expected, content, expected.Digest.String())
+	// copy content for referrers indexing
+	var buf bytes.Buffer
+	lr := limitReader(content, s.repo.MaxMetadataBytes)
+	tr := io.TeeReader(lr, &buf)
```
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply registry/remote/repository.go:907]: If referrers state is known to be supported, we don't need to buffer anything.

File: registry/remote/repository.go [hunk 3]
```diff
@@ -1051,6 +1090,113 @@ func (s *manifestStore) push(ctx context.Context, expected ocispec.Descriptor, c
...
+	if err != nil && !errors.Is(err, errdef.ErrNotFound) {
+		return err
+	}
+
+	updatedReferrers = append(updatedReferrers, desc)
+	indexDesc, index, err := generateReferrersIndex(updatedReferrers)
+	if err != nil {
+		return fmt.Errorf("failed to generate referrers index for referrers tag %s: %w", referrersTag, err)
+	}
+	return s.push(ctx, indexDesc, bytes.NewReader(index), referrersTag)
```
Comment (shizhMSFT) [inline registry/remote/repository.go:1160]: After pushing, we need to delete previous index.

File: registry/remote/repository.go [hunk 4]
```diff
@@ -1190,3 +1336,24 @@ func verifyContentDigest(resp *http.Response, expected digest.Digest) error {
 
 	return nil
 }
+
+// generateReferrersIndex generates an image index containing the given
+// referrers list in its manifests field.
+func generateReferrersIndex(referrers []ocispec.Descriptor) (ocispec.Descriptor, []byte, error) {
+	if referrers == nil {
+		referrers = []ocispec.Descriptor{} // make it an empty array to prevent potential server-side bugs
+	}
```
Comment (shizhMSFT) [inline registry/remote/repository.go:1345]: If `len(referrers) == 0`, we don't even need to generate the index. Just let it go 🌬️🧊

File: registry/remote/repository.go [hunk 5]
```diff
@@ -1051,6 +1090,113 @@ func (s *manifestStore) push(ctx context.Context, expected ocispec.Descriptor, c
...
+	if yes {
+		// referrers API is available, no client-side indexing needed
+		return nil
+	}
+
+	// there can be multiple go-routines updating the referrers tag concurrently
+	referrersTag := buildReferrersTag(subject)
+	s.repo.lockReferrersTag(referrersTag)
+	defer s.repo.unlockReferrersTag(referrersTag)
```
Comment (shizhMSFT) [inline registry/remote/repository.go:1139]: There is an advanced concurrency technique to reduce multiple index re-generate event into one.

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: registry/remote/repository_test.go [hunk 6]
```diff
@@ -2767,6 +2768,377 @@ func Test_ManifestStore_Push(t *testing.T) {
...
+	err = repo.Push(ctx, manifestDesc, bytes.NewReader(manifestJSON))
+	if err != nil {
+		t.Fatalf("Manifests.Push() error = %v", err)
+	}
+	if !bytes.Equal(gotManifest, manifestJSON) {
+		t.Errorf("Manifests.Push() = %v, want %v", string(gotManifest), string(manifestJSON))
+	}
+}
+
+func Test_ManifestStore_Push_ReferrersAPIUnAvailable(t *testing.T) {
```
Comment (shizhMSFT) [inline registry/remote/repository_test.go:2875]: nit:
```suggestion
func Test_ManifestStore_Push_ReferrersAPIUnavailable(t *testing.T) {
```

File: registry/remote/repository.go [hunk 7]
```diff
@@ -116,6 +119,9 @@ type Repository struct {
 	// referrersState represents that if the repository supports Referrers API.
 	// default: referrersStateUnknown
 	referrersState referrersState
+
+	// referrersTagLocks maps a referrers tag to a lock.
+	referrersTagLocks sync.Map // map[string]sync.Mutex
```
Comment (wangxiaoxuan273) [inline registry/remote/repository.go:124]: Is this a useful comment or commented code? `// map[string]sync.Mutex`

Comment (Wwwsylvia) [reply registry/remote/repository.go:124]: It is a comment to show the map structure.

File: registry/remote/repository.go [hunk 8]
```diff
@@ -1051,6 +1090,113 @@ func (s *manifestStore) push(ctx context.Context, expected ocispec.Descriptor, c
 	return verifyContentDigest(resp, expected.Digest)
 }
 
+// indexReferrersForPush indexes referrers for image or artifact manifest with the
```
Comment (qweeah) [inline registry/remote/repository.go:1093]: nit: width this line goes beyond 80

File: registry/remote/repository.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply registry/remote/repository.go]: optimized

File: registry/remote/repository.go (hunk 1 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go]: This makes sense.

File: registry/remote/repository.go (hunk 3 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go]: Right, otherwise there will be many junk manifests.

File: registry/remote/repository.go (hunk 4 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go:1384]: This is just a helper function, will handler the `len(referrers) == 0` logic outside of this function.

File: registry/remote/repository.go (hunk 5 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go:1160]: What is that? I'm eager to learn!🤩

File: registry/remote/repository_test.go (hunk 6 above)
Comment (Wwwsylvia) [reply registry/remote/repository_test.go]: fixed!

File: registry/remote/repository.go (hunk 8 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go]: fixed!

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

File: registry/remote/repository.go (hunk 5 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go:1165]: Will optimize this later.

File: registry/remote/repository.go [hunk 9]
```diff
@@ -1051,6 +1090,113 @@ func (s *manifestStore) push(ctx context.Context, expected ocispec.Descriptor, c
...
+// isReferrersAPIAvailable returns true if the Referrers API is available for r.
+func (r *Repository) isReferrersAPIAvailable(ctx context.Context, desc ocispec.Descriptor) (bool, error) {
+	switch r.loadReferrersState() {
+	case referrersStateSupported:
+		return true, nil
+	case referrersStateUnsupported:
+		return false, nil
+	}
+
+	// referrers state is unknown
```
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply registry/remote/repository.go:1207]: Will optimize this later.

**Pull request merged**